pkg/ccl/streamingccl/streamingest/stream_ingestion_job_test.go

```diff
@@ -529,3 +530,83 @@ func TestCutoverFractionProgressed(t *testing.T) {
 	require.Equal(t, int64(0), metrics.ReplicationCutoverProgress.Value())
 	require.True(t, progressUpdates > 1)
 }
+
+// TestCutoverCheckpointing asserts that cutover progress persists to the job
+// record and ensures the cutover job does not duplicate persisted work after
+// the job is paused after a few updates.
+func TestCutoverCheckpointing(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	progressUpdated := make(chan struct{})
+	var updateCount int
+	remainingSpanUpdates := make(map[string]struct{})
+	args := replicationtestutils.DefaultTenantStreamingClustersArgs
+	args.TestingKnobs = &sql.StreamingTestingKnobs{
+		CutoverProgressShouldUpdate:  func() bool { return true },
+		OverrideRevertRangeBatchSize: 1,
+		OnCutoverProgressUpdate: func(remainingSpansUpdate roachpb.Spans) {
+
+			// If checkpointing works properly, we expect no repeating remaining span updates.
+			_, ok := remainingSpanUpdates[remainingSpansUpdate.String()]
+			require.False(t, ok, fmt.Sprintf("repeated remaining span update %s", remainingSpansUpdate.String()))
+			remainingSpanUpdates[remainingSpansUpdate.String()] = struct{}{}
+
+			updateCount++
+			if updateCount == 3 {
+				close(progressUpdated)
+				// To ensure the job machinery pauses the job before the next update
+				time.Sleep(1 * time.Second)
```

Comment:
IIUC if this time is not enough then the test will flake? because we will unpause and finish the cutover? I think it means it will flake..
why not just wait on a channel signaled after pause is completed? I guess in the next call to the hook?

if you're keeping this as it, try it under stress+race?

Reply:
good point, it seems to pass stress-race but i changed this to channel just to be safe.

Reply:
good point, it seems to pass stress-race but i changed this to a channel just to be safe.